### DISCUSSION TITLE
Subscribing to arrays

### DISCUSSION THREAD

**Discussion by DavidMulder0** [pmndrs/valtio#575] [Q&A]
I understand why `subscribe(store.arrayOfValues, ...)` breaks when `store.arrayOfValues = [...]` is assigned. I also understand why `subscribeKey(store, 'arrayOfValues')` doesn't trigger when `store.arrayOfValues.push(...)` is used. To handle both interactions with `store.arrayOfValues` you thus get this 'crazy' dance of having to do the `subscribe` inside (and outside) your `subscribeKey` handler... which gets messy (the messy part is in particular the unsubscribe handling, do I need to unsubscribe? Now that I think about it I probably don't?). 

What I basically wanted to ask is: Am I doing this wrong? How are you guys handling this? Before looking at the implementation of `subscribeKey` I was expecting it to fire on anything happening 'inside' the key as well. Right now I am thinking of just creating my own `subscribeKey` handler with a flag whether or not to subscribe to the targeted object, but wanted to ask around first.

**Comment by dai-shi** [comment]
If you want to accept both `= [...]` and `.push()`, I'd recommend to subscribe to the root `store`. `subscribeKey` is meant for primitive values, and it's a simple helper function, and it doesn't add any performance benefit compared to `subscribe` from the proxy perspective (it can reduce calling callbacks.)

My preference for mutable array is to use `.splice(0, len, ...)`, and `subscribe(store.arrayOfValues)`, which is better in performance.
